lib/BC/Util.cpp

```diff
@@ -1684,6 +1698,12 @@ void CloneFunctionInto(llvm::Function *source_func, llvm::Function *dest_func,
 
 
       new_insts.push_back(new_inst);
+#else
+      // TODO(alex): Figure out how we can reproduce the above hack in LLVM 16...
```

Comment:
I still need to figure this out.

Reply:
Also, I would prefer to move as much of these changes as possible to `Compat/`.

Reply:
Oh actually, I just realised something. The whole point of that hack above is to get access to a private member. We could probably just do that one more time.
- Once to get a handle on the instruction list.
- Once to mutate the instruction like we were doing before.
Not sure how I feel about that though... 🤮

Reply:
Overall, this was terrible / evil stuff. The fundamental problem was that I was too lazy to implement cloning instructions, so I wanted to reuse `->clone` and that had issues.

Reply:
We could just not port the hack.

Reply:
@pgoodman Hmm, what would that look like? I'm imagining some giant switch that covers every instruction type in LLVM and then recreates it in the target module. Is that what you had in mind?

Reply:
I think that's what LLVM does. The better question is: is it worth it to support cloning from one llvm context into another llvm context. If it's not worth it, then we can just error on the code. **My opinion is that supporting this use case is not worth the effort.**

It used to be in IRENE2 that I was trying to support different context cloning because I was sharing results with a different thread, and llvm contexts and other methods are not thread-safe. I ended up running up against a wall with the cloning approach, and abandoned it in favour of a serializing to IR to a string, then deserializing in the new LLVMContext, which was much safer and worked correctly, and so this stuff in Remill largely became dead code without a usecase.

Reply:
Ok cool, that seems like a good idea. At the moment, the Remill test suite uses this clone function helper and Anvill uses it directly. I'll try to see if I can write a helper that serializes to a string (shouldn't be hard considering that we've got one that writes it to a file) and then try to replace those usages.

Reply:
Does the test suite actually alter the LLVM context?

Reply:
Yes. It moves functions into their own module so they can be executed: https://github.com/lifting-bits/remill/blob/master/tests/PPC/TestLifting.cpp#L203

Reply:
That looks like it would reuse the same context? That is, there is no explicit passing of a new context, so presumably it shares the same LLVM context with the input module.

Reply:
Oh sorry, I thought you meant module not context. What's the consequence of that though? Seems like I'd still have trouble moving the function across modules.

Reply:
I don't think there are any issues with cross-module cloning, as long as the llvm context is shared between them. The instruction cloning is find cross-module as long as it's not cross-context.

Reply:
[bf59658](https://github.com/lifting-bits/remill/pull/665/commits/bf59658d796a8e6870adc7c6cefb6c61a37c3384)

Gotcha, thanks for explaining. I think we can just remove the part where we recontextualize the type and leave everything as is.